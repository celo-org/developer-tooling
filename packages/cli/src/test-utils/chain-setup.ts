import { ContractKit } from '@celo/contractkit'
import { mineBlocks } from '@celo/dev-utils/lib/ganache-test'
import { addressToPublicKey } from '@celo/utils/lib/signatureUtils'
import BigNumber from 'bignumber.js'
import Web3 from 'web3'

export const GANACHE_EPOCH_SIZE = 100
export const MIN_LOCKED_CELO_VALUE = new BigNumber(Web3.utils.toWei('10000', 'ether')) // 10k CELO

const GROUP_COMMISION = new BigNumber(0.1)
const BLS_PUBLIC_KEY =
  '0x4fa3f67fc913878b068d1fa1cdddc54913d3bf988dbe5a36a20fa888f20d4894c408a6773f3d7bde11154f2a3076b700d345a42fd25a0e5e83f4db5586ac7979ac2053cd95d8f2efd3e959571ceccaa743e02cf4be3f5d7aaddb0b06fc9aff00'
const BLS_POP =
  '0xcdb77255037eb68897cd487fdd85388cbda448f617f874449d4b11588b0b7ad8ddc20d9bb450b513bb35664ea3923900'

export const registerAccount = async (kit: ContractKit, address: string) => {
  const accounts = await kit.contracts.getAccounts()

  if (!(await accounts.isAccount(address))) {
    await accounts.createAccount().sendAndWaitForReceipt({ from: address })
  }
}

export const registerAccountWithLockedGold = async (
  kit: ContractKit,
  address: string,
  value: string = MIN_LOCKED_CELO_VALUE.toFixed()
) => {
  await registerAccount(kit, address)

  const lockedGold = await kit.contracts.getLockedGold()

  await lockedGold.lock().sendAndWaitForReceipt({ from: address, value })
}

export const setupGroup = async (
  kit: ContractKit,
  groupAccount: string,
  groupCommission = GROUP_COMMISION
) => {
  await registerAccountWithLockedGold(
    kit,
    groupAccount,
    new BigNumber(MIN_LOCKED_CELO_VALUE).toFixed()
  )

  const validators = await kit.contracts.getValidators()

  await (
    await validators.registerValidatorGroup(groupCommission)
  ).sendAndWaitForReceipt({
    from: groupAccount,
  })
}

const setupValidator = async (kit: ContractKit, validatorAccount: string) => {
  await registerAccountWithLockedGold(kit, validatorAccount)

  const ecdsaPublicKey = await addressToPublicKey(validatorAccount, kit.connection.sign)
  const validators = await kit.contracts.getValidators()

  await validators
    .registerValidator(ecdsaPublicKey, BLS_PUBLIC_KEY, BLS_POP)
    .sendAndWaitForReceipt({
      from: validatorAccount,
    })
}

export const setupGroupAndAffiliateValidator = async (
  kit: ContractKit,
  groupAccount: string,
  validatorAccount: string
) => {
  await setupGroup(kit, groupAccount)
  await setupValidator(kit, validatorAccount)

  const validators = await kit.contracts.getValidators()

  await validators.affiliate(groupAccount).sendAndWaitForReceipt({ from: validatorAccount })

  await (
    await validators.addMember(groupAccount, validatorAccount)
  ).sendAndWaitForReceipt({
    from: groupAccount,
  })
}

export const voteForGroupFrom = async (
  kit: ContractKit,
  fromAddress: string,
  groupAddress: string,
  amount: BigNumber
) => {
  const election = await kit.contracts.getElection()

  await (await election.vote(groupAddress, amount)).sendAndWaitForReceipt({ from: fromAddress })
}

export const voteForGroupFromAndActivateVotes = async (
  kit: ContractKit,
  fromAddress: string,
  groupAddress: string,
  amount: BigNumber
) => {
  await voteForGroupFrom(kit, fromAddress, groupAddress, amount)

  await mineEpoch(kit)

  const election = await kit.contracts.getElection()

  const txos = await election.activate(fromAddress, false)

  await Promise.all(txos.map((txo) => txo.sendAndWaitForReceipt({ from: fromAddress })))
}

export const mineEpoch = async (kit: ContractKit) => {
  await mineBlocks(100, kit.web3)
}
