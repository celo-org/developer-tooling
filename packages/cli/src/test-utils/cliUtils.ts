import { Interfaces } from '@oclif/core'
import { BaseCommand } from '../base'

type AbstractConstructor<T> = new (...args: any[]) => T
interface Runner extends AbstractConstructor<BaseCommand> {
  run: typeof BaseCommand.run
}

export async function testLocally(
  command: Runner,
  argv: string[],
  config?: Interfaces.LoadOptions
) {
  const extendedArgv = [...argv, '--node', 'local']
  return command.run(extendedArgv, config)
}

// Removes font-formatting ANSI codes (colors/styles)
export const stripAnsiCodes = (text: string): string => {
  return text.replace(/\u001b\[.*?m/g, '')
}

export function stripAnsiCodesFromNestedArray(arrays: Array<string[]>) {
  return arrays.map((level0) => level0.map((level1) => stripAnsiCodes(level1)))
}

export const LONG_TIMEOUT_MS = 10 * 1000
export const EXTRA_LONG_TIMEOUT_MS = 20 * 1000
