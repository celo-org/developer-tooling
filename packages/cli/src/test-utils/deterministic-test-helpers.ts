import { HttpRpcCaller } from '@celo/connect'

/**
 * Mock gas prices to be deterministic across environments
 */
export const mockDeterministicGas = () => {
  return jest.spyOn(HttpRpcCaller.prototype, 'call').mockImplementation(async (method, _args) => {
    // Set deterministic gas prices that match CI environment
    if (method === 'eth_gasPrice') {
      return {
        result: '0x5d21dba00', // 25000000000 - deterministic gas price
        id: 1,
        jsonrpc: '2.0',
      }
    }
    if (method === 'eth_maxPriorityFeePerGas') {
      return {
        result: '0x4e3b29200', // 20000000000 - deterministic priority fee
        id: 1,
        jsonrpc: '2.0',
      }
    }
    if (method === 'eth_feeHistory') {
      return {
        result: {
          baseFeePerGas: ['0x5d21dba00'], // Same as gas price
          gasUsedRatio: [0.5],
          reward: [['0x4e3b29200']],
        },
        id: 1,
        jsonrpc: '2.0',
      }
    }
    // For other methods, call through to original implementation
    return HttpRpcCaller.prototype.call.call(this, method, _args)
  })
}

/**
 * Mock balance queries to return deterministic values
 */
export const mockDeterministicBalance = (expectedBalance: string) => {
  return jest.spyOn(HttpRpcCaller.prototype, 'call').mockImplementation(async (method, args) => {
    if (method === 'eth_getBalance') {
      return {
        result: expectedBalance, // Use the expected balance from CI snapshots
        id: 1,
        jsonrpc: '2.0',
      }
    }
    // For other methods, call through to original implementation
    return HttpRpcCaller.prototype.call.call(this, method, args)
  })
}