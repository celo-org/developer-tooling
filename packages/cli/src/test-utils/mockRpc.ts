import { HttpRpcCaller } from '@celo/connect'
import { RpcErrorCode } from 'viem'

export const mockRpc = () =>
  jest.spyOn(HttpRpcCaller.prototype, 'call').mockImplementation(async (method, _args) => {
    if (method === 'eth_maxPriorityFeePerGas') {
      return {
        result: '20000',
        id: 1,
        jsonrpc: '2.0',
      }
    }
    if (method === 'eth_gasPrice') {
      return {
        result: '30000',
        id: 1,
        jsonrpc: '2.0',
      }
    }
    return {
      result: 0,
      id: Math.random(),
      jsonrpc: '2.0',
    }
  })

const actualFetch = global.fetch
export const mockRpcFetch = ({
  method,
  result,
  error,
}: {
  method: string | string[]
  result?: any
  error?: { code: RpcErrorCode; message?: string }
}) => {
  const fetchMock = jest.fn(async (...args) => {
    if (args[1]?.body) {
      const body = JSON.parse(args[1].body.toString())
      if (typeof method === 'string' ? body.method === method : method.includes(body.method)) {
        return {
          ok: !error,
          headers: new Map([['Content-Type', 'application/json']]),
          json: () => {
            if (error) {
              return Promise.resolve({ error })
            } else {
              return Promise.resolve({ result, id: 1, jsonrpc: '2.0' })
            }
          },
        }
      }
    }
    // @ts-expect-error
    return actualFetch(...args)
  })
  // @ts-expect-error
  global.fetch = fetchMock

  return function restoreFetch() {
    global.fetch = actualFetch
  }
}
