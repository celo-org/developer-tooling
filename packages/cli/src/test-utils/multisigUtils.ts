import { multiSigABI, proxyABI } from '@celo/abis'
import { StrongAddress } from '@celo/base'
import { ContractKit } from '@celo/contractkit'
import { multiSigBytecode, proxyBytecode } from './constants'

export async function createMultisig(
  kit: ContractKit,
  owners: StrongAddress[],
  requiredSignatures: number,
  requiredInternalSignatures: number
): Promise<StrongAddress> {
  const accounts = (await kit.web3.eth.getAccounts()) as StrongAddress[]
  kit.defaultAccount = accounts[0]

  // Deploy Proxy contract
  const proxyDeploymentTx = await kit.sendTransaction({
    data: proxyBytecode,
  })
  const { contractAddress: proxyAddress } = await proxyDeploymentTx.waitReceipt()
  // Deploy MultiSig contract
  const multisigDeploymentTx = await kit.sendTransaction({
    data: multiSigBytecode,
  })
  const { contractAddress: multiSigAddress } = await multisigDeploymentTx.waitReceipt()

  // Configure and initialize MultiSig
  const initializerAbi = multiSigABI.find(
    (abi) => abi.type === 'function' && abi.name === 'initialize'
  )
  const proxy = new kit.web3.eth.Contract(proxyABI as any, proxyAddress)
  const baseFee = await kit.web3.eth.getBlock('latest').then((block: any) => block.baseFeePerGas)
  const priorityFee = kit.web3.utils.toWei('2', 'gwei')
  const initMethod = proxy.methods._setAndInitializeImplementation
  const callData = kit.web3.eth.abi.encodeFunctionCall(initializerAbi as any, [
    owners as any,
    requiredSignatures as any,
    requiredInternalSignatures as any,
  ])
  const initTx = initMethod(multiSigAddress, callData)
  await initTx.send({
    from: kit.defaultAccount,
    gas: await initTx.estimateGas({ from: kit.defaultAccount }),
    maxPriorityFeePerGas: priorityFee,
    maxFeePerGas: (BigInt(baseFee) + BigInt(priorityFee)).toString(),
  })
  const transferOwnershipMethod = proxy.methods._transferOwnership
  const changeOwnerTx = transferOwnershipMethod(proxyAddress)
  await changeOwnerTx.send({
    from: kit.defaultAccount,
    gas: await changeOwnerTx.estimateGas({ from: kit.defaultAccount }),
    maxPriorityFeePerGas: priorityFee,
    maxFeePerGas: (BigInt(baseFee) + BigInt(priorityFee)).toString(),
  })

  return proxyAddress as StrongAddress
}
