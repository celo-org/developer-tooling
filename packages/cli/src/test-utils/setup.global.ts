import baseSetup from '@celo/dev-utils/lib/ganache-setup'
// Has to import the matchers somewhere so that typescript knows the matchers have been made available
import _unused from '@celo/dev-utils/lib/matchers'
import * as path from 'path'

// Warning: There should be an unused import of '@celo/dev-utils/lib/matchers' above.
// If there is not, then your editor probably deleted it automatically.

export default async function globalSetup() {
  console.log('\nstarting ganache...')
  const chainDataPath = path.join(path.dirname(require.resolve('@celo/celo-devchain')), '../chains')
  // v X refers to core contract release X
  await baseSetup(path.resolve(chainDataPath), 'v11.tar.gz', {
    from_targz: true,
  })
  console.log('\n ganache started...')
  // it is necessary to disabled oclif integration with ts-node as
  // together it leads to a silent signit error and exit when tsconfk is loaded.
  // @ts-ignore - because global this doesnt have oclif property
  global.oclif = { tsnodeEnabled: false }
}
