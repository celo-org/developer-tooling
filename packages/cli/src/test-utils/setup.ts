jest.mock('@ledgerhq/hw-transport-node-hid', () => {
  return {
    default: {
      open: jest.fn(() => {
        return {
          send: jest.fn(() => new Promise(() => {})),
          decorateAppAPIMethods: jest.fn(),
          close: jest.fn(),
        }
      }),
    },
  }
})
