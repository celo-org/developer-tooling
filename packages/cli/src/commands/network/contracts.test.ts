import { testWithGanache } from '@celo/dev-utils/lib/ganache-test'
import write from '@oclif/core/lib/cli-ux/write'
import { testLocally } from '../../test-utils/cliUtils'
import Contracts from './contracts'
process.env.NO_SYNCCHECK = 'true'

testWithGanache('network:contracts', () => {
  test('runs', async () => {
    const spy = jest.spyOn(write, 'stdout')
    await testLocally(Contracts, ['--output', 'json'])
    expect(spy.mock.calls).toMatchSnapshot()
  })
})
