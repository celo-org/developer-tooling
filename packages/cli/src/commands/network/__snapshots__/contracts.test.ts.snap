// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`network:contracts runs 1`] = `
[
  [
    "[
  {
    "contract": "Accounts",
    "proxy": "0xa9a65D631f8c8577f543B64B35909030C84676A2",
    "implementation": "0x038C860fd0d598B3DF5577B466c8b0a074867f56",
    "version": "1.1.4.1"
  },
  {
    "contract": "Attestations",
    "proxy": "0x54a582397Ace3F9a6F5A28cE6Dc50473E4A9d375",
    "implementation": "0xc650825Ca03EeB83d0d075c2737022cBb01F5799",
    "version": "1.2.0.0"
  },
  {
    "contract": "BlockchainParameters",
    "proxy": "0xd418d14C9b8aA740E8D18C6dfda3af6f023A27A5",
    "implementation": "0x4CF20BEf2FBe159601939bE49B760457B11C1d52",
    "version": "1.3.0.0"
  },
  {
    "contract": "DoubleSigningSlasher",
    "proxy": "0x175Ac784563DE645647C6350F0CFc577Dcc7eE5b",
    "implementation": "0xA47aA5fCecBfb374E79144F05fDf91D0F50fA351",
    "version": "1.1.1.0"
  },
  {
    "contract": "DowntimeSlasher",
    "proxy": "0x0C5173a51e26b29d6126C686756FB9FBEf71f762",
    "implementation": "0x33a89cA59bc3A9221456ffE461de17C8deF24684",
    "version": "2.0.0.0"
  },
  {
    "contract": "Election",
    "proxy": "0xcBAe15A320F56Fc9ad6c8319D55Be1FeF0750070",
    "implementation": "0x51815ebd3C922B215b0d0B9b41a4Daa0dB70b841",
    "version": "1.1.3.0"
  },
  {
    "contract": "EpochRewards",
    "proxy": "0x537732b0bAC4F2A1a97b1a60cEa0df5C3f0DEF16",
    "implementation": "0x1E28a5fB7B112291F088bBB8ab693D2214DB7895",
    "version": "1.1.1.0"
  },
  {
    "contract": "Escrow",
    "proxy": "0xE8593eFb73423D86e5F337A46B3450eEec1027c6",
    "implementation": "0xc51B43db0Cea40E36207993F0aB1883E7A865417",
    "version": "1.2.0.0"
  },
  {
    "contract": "FederatedAttestations",
    "proxy": "0x8C3d218745c360A399285c0176714530Ce0af3B1",
    "implementation": "0x9d0e84A58Bb5B9b94F5c34fD4f310ed3F41A2D69",
    "version": "1.1.0.0"
  },
  {
    "contract": "FeeCurrencyWhitelist",
    "proxy": "0xE86bB98fcF9BFf3512C74589B78Fb168200CC546",
    "implementation": "0xDc688D29394a3f1E6f1E5100862776691afAf3d2",
    "version": "NONE"
  },
  {
    "contract": "FeeHandler",
    "proxy": "0x176097114008844bCb75Ea58998B7f8303b4C2d4",
    "implementation": "0x123d0530A2a37FA4d4efA3999e79203BAf5bE517",
    "version": "1.1.0.0"
  },
  {
    "contract": "Freezer",
    "proxy": "0xcFC18CEc799fBD1793B5C43E773C98D4d61Cc2dB",
    "implementation": "0xF22469F31527adc53284441bae1665A7b9214DBA",
    "version": "NONE"
  },
  {
    "contract": "GasPriceMinimum",
    "proxy": "0x038F9B392Fb9A9676DbAddF78EA5fdbf6C7d9710",
    "implementation": "0x371b13d97f4bF77d724E78c16B7dC74099f40e84",
    "version": "1.2.0.1"
  },
  {
    "contract": "GoldToken",
    "proxy": "0x04B5dAdd2c0D6a261bfafBc964E0cAc48585dEF3",
    "implementation": "0x8726C7414ac023D23348326B47AF3205185Fd035",
    "version": "1.1.2.0"
  },
  {
    "contract": "Governance",
    "proxy": "0x82E9A894F6b752E5c1058ea69898Ff5D3EBA3939",
    "implementation": "0xc98E260cBF7041cc2C42D7e055fb4350DF22dd68",
    "version": "1.4.1.0"
  },
  {
    "contract": "LockedGold",
    "proxy": "0x4A86aD5f263260f24483Df2B1B3a23ea4788B6AB",
    "implementation": "0x0221652D4306F6b3CB7B23E43C3e25D8F9a142Ca",
    "version": "1.1.4.0"
  },
  {
    "contract": "MentoFeeHandlerSeller",
    "proxy": "0x98942DE2B8eaC1c264e41cA981b82032762d72a4",
    "implementation": "0x8501835ac2B234147F766bB6e6423D6Ef50A9247",
    "version": "1.1.0.0"
  },
  {
    "contract": "OdisPayments",
    "proxy": "0x347BfD5F56F04210603c9cC7c6c1f88E9E5c9676",
    "implementation": "0x05E0c172eB6244A564B0D9e59Ce63074584a1e7A",
    "version": "1.1.0.0"
  },
  {
    "contract": "Random",
    "proxy": "0xd2dBF3250a764eaaa94fa0C84ED87C0Edc8eD04E",
    "implementation": "0x39c3Fc9F4D8430af2713306CE80C584752d9e1C7",
    "version": "1.1.1.0"
  },
  {
    "contract": "Registry",
    "proxy": "0x000000000000000000000000000000000000ce10",
    "implementation": "0x07f96Aa816C1F244CbC6ef114bB2b023Ba54a2EB",
    "version": "NONE"
  },
  {
    "contract": "Reserve",
    "proxy": "0xCA9717a4A6e8009B3518648C9f3E7676255471A1",
    "implementation": "0x4586649629F699f9A4B61D0e962DC3c9025Fe488",
    "version": "1.1.2.2"
  },
  {
    "contract": "SortedOracles",
    "proxy": "0x4D3d5c850Dd5bD9D6F4AdDA3DD039a3C8054CA29",
    "implementation": "0xA31E64EA55B9B6Bbb9d6A676738e9A5b23149f84",
    "version": "1.1.3.0"
  },
  {
    "contract": "StableToken",
    "proxy": "0x5315e44798395d4a952530d131249fE00f554565",
    "implementation": "0xDFF540fE764855D3175DcfAe9d91AE8aEE5C6D6F",
    "version": "NONE"
  },
  {
    "contract": "StableTokenBRL",
    "proxy": "0x965D352283a3C8A016b9BBbC9bf6306665d495E7",
    "implementation": "0x9a1df498af690a7EB43E10A28AB51345a3A33F75",
    "version": "NONE"
  },
  {
    "contract": "StableTokenEUR",
    "proxy": "0xdD66C23e07b4D6925b6089b5Fe6fc9E62941aFE8",
    "implementation": "0x4609e0ED27A8BAAc57b753D36a5D2971915588f9",
    "version": "NONE"
  },
  {
    "contract": "UniswapFeeHandlerSeller",
    "proxy": "0x9EF3dACB3367d7741099b5607A7a93E99a119F7b",
    "implementation": "0x0F3723fB2a7C147357868702164f9beEf54E1E85",
    "version": "1.1.0.0"
  },
  {
    "contract": "Validators",
    "proxy": "0x6AC7189BD267c81c55d7d7d0321f9B23639cB9db",
    "implementation": "0x3E809c563c15a295E832e37053798DdC8d6C8dab",
    "version": "1.2.0.5"
  }
]
",
  ],
]
`;
